### ISSUE TITLE
Typecheck default values if unusued

### ISSUE THREAD

**Issue by pkorotkov** [crystal-lang/crystal#2636]
This code triggers an expected error

``` cr
class A(T)
  getter values

  def initialize(@values : Array(T) = Int)
  end
end

a = A(String).new

p typeof(a.values)
```

[#link 1](https://play.crystal-lang.org/#/r/za8)

but the following snippet compiles and runs just fine contrary to my expectations

``` cr
class A(T)
  getter values

  def initialize(@values : Array(T) = Int)
  end
end

a = A.new(values = ["One", "Two"])

p typeof(a.values)
```

[#link 2](https://play.crystal-lang.org/#/r/za9)

**Comment by jhass** [comment]
Reduced

```cr
class A
  def initialize(@values : Int32 = true)
  end
end

a = A.new
```
vs
```cr
class A
  def initialize(@values : Int32 = true)
  end
end

a = A.new(1)
```

**Comment by asterite** [comment]
Not sure this is a bug. Right now you can do:

```crystal
# You can only pass a String, but the default value is nil
def foo(value : String = nil)
end

foo # OK
foo "hello" # OK
foo nil # Error
```

In that way you can restrict an argument to a type but use a default value of another type.

**Comment by jhass** [comment]
That's just even more inconsistent, isn't it?

**Comment by asterite** [comment]
I was thinking about having `nil` there to know that a value that's not possible to pass was received in the method, so you know no argument has been passed. This could also be useful if you use the null object pattern. But yeah, it's probably inconsistent. I'll try to fix this, though the error will happen if you invoke the method (as usual), so the second snippet in the first comment will still compile fine (so the end result will be more or less the one we have right now)

**Comment by bcardiff** [comment]
IMO it is a bug. It will force to implementor some `T?` for nillable arguments but I think is the right thing to do.

Maybe there will be a tricky part when the method involves type variables. I was trying some snippets but might reach a separate issue here:

https://play.crystal-lang.org/#/r/zbd
```crystal
def foo(o : K, h : Hash(K, V) = { "foo" => "bar" })
  h.keys
end

pp foo(7, {1 => 2}) # => [1]
pp foo("sdf") # => ["foo"]
pp foo(8) # => ["foo"] 
```

**Comment by asterite** [comment]
Part of this was fixed in #3834

The original title of this issue is "Typecheck default values if unusued", which basically means that this should give a compile error:

```cr
def foo(x : Int32 = "hello")
end
```

That is, it should give a compile error even if we don't invoke `foo`. I'm not sure about that. Right now we only type-check things on invocation. For example this works fine in Ruby:

```rb
def foo(x = unknown)
end

foo(1)
```

But if we don't pass an argument we get an error. Well, it's the same in Crystal here, and I think that's OK (having to type check all methods even if they are not used will slow down compilation, and I'm not sure it's even possible in all cases)
